lib/codegen-meta/src/cdsl/types.rs

```diff
@@ -0,0 +1,467 @@
+        }
+    }
+
+    /// Return the number of bits in a lane.
+    pub fn lane_bits(&self) -> u64 {
+        self.bits
+    }
+
+    /// Get the name of this lane type.
+    pub fn name(&self) -> String {
+        match self.tag {
+            LaneTypeTag::BoolType(_) => format!("b{}", self.bits),
+            LaneTypeTag::FloatType(_) => format!("f{}", self.bits),
+            LaneTypeTag::IntType(_) => format!("i{}", self.bits),
+        }
+    }
+
+    /// Find the unique number associated with this lane type.
+    pub fn number(&self) -> u8 {
+        match self.tag {
+            LaneTypeTag::BoolType(b) => b.number(),
+            LaneTypeTag::FloatType(f) => f.number(),
+            LaneTypeTag::IntType(i) => i.number(),
+        }
+    }
+}
+
+impl fmt::Debug for LaneType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let inner_msg = format!("bits={}", self.lane_bits());
+        write!(
+            f,
+            "{}",
+            match self.tag {
+                LaneTypeTag::BoolType(_) => format!("BoolType({})", inner_msg),
+                LaneTypeTag::FloatType(_) => format!("FloatType({})", inner_msg),
+                LaneTypeTag::IntType(_) => format!("IntType({})", inner_msg),
+            }
+        )
+    }
+}
+
+/// Create a LaneType from a given bool variant.
+impl From<base_types::Bool> for LaneType {
+    fn from(b: base_types::Bool) -> Self {
+        let bits = b as u64;
+        let tag = LaneTypeTag::BoolType(b);
+        Self { bits, tag }
+    }
+}
+
+/// Create a LaneType from a given float variant.
+impl From<base_types::Float> for LaneType {
+    fn from(f: base_types::Float) -> Self {
+        let bits = f as u64;
+        let tag = LaneTypeTag::FloatType(f);
+        Self { bits, tag }
+    }
+}
+
+/// Create a LaneType from a given int variant.
+impl From<base_types::Int> for LaneType {
+    fn from(i: base_types::Int) -> Self {
+        let bits = i as u64;
+        let tag = LaneTypeTag::IntType(i);
+        Self { bits, tag }
+    }
+}
+
+/// Tags used to specify the kinds of elements in a lane type.
+#[derive(Debug, Clone, Copy)]
+pub enum LaneTypeTag {
+    BoolType(base_types::Bool),
+    FloatType(base_types::Float),
+    IntType(base_types::Int),
+}
+
+/// An iterator for different lane types.
+pub struct LaneTypeIterator {
+    bool_iter: base_types::BoolIterator,
+    int_iter: base_types::IntIterator,
+    float_iter: base_types::FloatIterator,
+}
+
+impl LaneTypeIterator {
+    /// Create a new lane type iterator.
+    fn new() -> Self {
+        Self {
+            bool_iter: base_types::BoolIterator::new(),
+            int_iter: base_types::IntIterator::new(),
+            float_iter: base_types::FloatIterator::new(),
+        }
+    }
+}
+
+impl Iterator for LaneTypeIterator {
+    type Item = LaneType;
+    fn next(&mut self) -> Option<Self::Item> {
+        if let b @ Some(_) = self.bool_iter.next() {
+            b.map(LaneType::from)
```

Comment:
Since the method returns an option, this could also rewritten as:

```rust
      if let Some(b) = self.bool_iter.next() {
            Some(LaneType::from(b))
```

I can write it that way, so it isn't quite as opaque :)